Lib/tarfile.py

```diff
@@ -2349,6 +2350,8 @@ def next(self):
                     raise ReadError(str(e)) from None
             except SubsequentHeaderError as e:
                 raise ReadError(str(e)) from None
+            except zlib.error as e:
+                raise ExtractError('archive may be corrupted or invalid') from e
```

Comment:
1. This should be a `ReadError` instead. `ExtractError` would be a new kind of error that is currently not raised by the `next()` function, making existing code invalid. `ReadError` is raised by other branches already so using it is alright as it's got better backwards compatibility.
2. While I like the idea to fix leaking implementation details here and to return a tarfile Error instead of an underlying exception, you need to catch a bare `Exception` here and try importing zlib from there. If you're already doing that, it would be great if you checked what happens for invalid bz2 and lzma files, too.

Reply:
> it would be great if you checked what happens for invalid bz2 and lzma files, too

The example that was originally posted on the bpo was actually not compressed; it's just a tar archive. I believe that passing invalid tar archives is the key to reproducing this bug, not invalid compressed data. I tried throwing many flavors of invalid compressed data at `tarfile.open`, and it seems to do the right thing.

I invite you to view my spaghetti code, if you dare
https://gist.github.com/jdevries3133/acbb5ba2a19093d3bcc214733ef85e5a

I think the next step will be for me to corrupt uncompressed archives, then compress them, then pass those into `tarfile.open`; that's the only thing I have not done yet. However, I expect it will behave the same as the original bpo example.